### PR TITLE
[perf] client/server experiments

### PR DESCRIPTION
**Pull request by pruivo**
compute hash code lazily instead of when unmarshalling key/values

### CODE REVIEW

**Comment by pruivo** [comment]
run performance tests please
cs dist

**Comment by pruivo** [comment]
run performance tests please
cs dist write
library dist write

File: core/src/main/java/org/infinispan/interceptors/distribution/TriangleDistributionInterceptor.java [hunk 1]
```diff
@@ -324,37 +356,36 @@ private Object primaryOwnerWrite(InvocationContext context, DataWriteCommand com
...
    }
 
-   private void logCommandSequence(CommandInvocationId id, int segment, long sequence) {
-         log.tracef("Command %s got sequence %s for segment %s", id, sequence, segment);
+   private void sendToBackup(DataWriteCommand command, Collection<Address> backups, int segmentId) {
+      long sequenceNumber = triangleOrderManager.next(segmentId, command.getTopologyId());
+      BackupWriteRcpCommand backupWriteRcpCommand = commandsFactory.buildBackupWriteRcpCommand(command);
+      backupWriteRcpCommand.setSequence(sequenceNumber);
+      if (trace) {
+         logCommandSequence(command.getCommandInvocationId(), segmentId, sequenceNumber, backups);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/distribution/TriangleDistributionInterceptor.java]: What does this pattern buy us? Doesn't the JIT eliminate everything inside `if (trace)` anyway?

File: core/src/main/java/org/infinispan/util/ByteString.java [hunk 2]
```diff
@@ -39,7 +38,7 @@ public static ByteString emptyString() {
 
    @Override
    public int hashCode() {
-      return hash;
+      return hash == 0 ? (hash = Arrays.hashCode(b)) : hash;
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/ByteString.java]: Do we ever create a `ByteString` without looking it up in a map of caches?

File: commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java [hunk 3]
```diff
@@ -74,7 +73,7 @@ public boolean equalsWrappedBytes(WrappedBytes other) {
 
    @Override
    public int hashCode() {
-      return hashCode;
+      return hashCode == 0 ? (hashCode = Arrays.hashCode(bytes)) : hashCode;
```
Comment (danberindei) [inline commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java]: Maybe we shouldn't cache the hash code at all? `MurmurHash3.hash(Object)` doesn't use it anyway...

File: core/src/main/java/org/infinispan/interceptors/distribution/TriangleDistributionInterceptor.java (hunk 1 above)
Comment (pruivo) [inline core/src/main/java/org/infinispan/interceptors/distribution/TriangleDistributionInterceptor.java]: > What does this pattern buy us?

headaches...

File: core/src/main/java/org/infinispan/util/ByteString.java (hunk 2 above)
Comment (pruivo) [inline core/src/main/java/org/infinispan/util/ByteString.java]: no. the first thing we do when we create it, is to put it in a map or check it in a map :P

File: commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java (hunk 3 above)
Comment (pruivo) [inline commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java]: AFAIK, it is only used to put keys in DataContainer

File: core/src/main/java/org/infinispan/interceptors/distribution/TriangleDistributionInterceptor.java (hunk 1 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/distribution/TriangleDistributionInterceptor.java]: :D

File: commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java (hunk 3 above)
Comment (danberindei) [inline commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java]: I guess it's used for reading as well, so a write operation would benefit from the cached hash code.

OTOH with compressed oops removing the field will remove 8 bytes from `WrappedByteArray`, and keys aren't supposed to be that long anyway, so I would still remove the field.

**Pull request closed without merging**